pkg/kubectl/rollback.go

```diff
@@ -195,3 +207,108 @@ func simpleDryRun(deployment *extensions.Deployment, c clientset.Interface, toRe
 	printersinternal.DescribePodTemplate(internalTemplate, w)
 	return buf.String(), nil
 }
+
+type DaemonSetRollbacker struct {
+	c clientset.Interface
+}
+
+func (r *DaemonSetRollbacker) Rollback(obj runtime.Object, updatedAnnotations map[string]string, toRevision int64, dryRun bool) (string, error) {
+	if toRevision < 0 {
+		return "", revisionNotFoundErr(toRevision)
+	}
+
+	ds, ok := obj.(*extensions.DaemonSet)
+	if !ok {
+		return "", fmt.Errorf("passed object is not a DaemonSet: %#v", obj)
+	}
+	allHistory, err := controlledHistories(r.c, ds)
+	if err != nil {
+		return "", fmt.Errorf("unable to find history controlled by DaemonSet %s: %v", ds.Name, err)
+	}
+
+	if toRevision == 0 && len(allHistory) <= 1 {
+		return "", fmt.Errorf("no last revision to roll back to")
+	}
+
+	// Find the history to rollback to
+	var toHistory *appsv1beta1.ControllerRevision
+	if toRevision == 0 {
+		// If toRevision == 0, find the latest revision (2nd max)
+		sort.Sort(historiesByRevision(allHistory))
+		toHistory = allHistory[len(allHistory)-2]
+	} else {
+		for _, h := range allHistory {
+			if h.Revision == toRevision {
+				// If toRevision != 0, find the history with matching revision
+				toHistory = h
+				break
+			}
+		}
+	}
+	if toHistory == nil {
+		return "", revisionNotFoundErr(toRevision)
+	}
+
+	// Get the template of the history to rollback to
+	toTemplate, err := getInternalTemplate(toHistory)
+	if err != nil {
+		return "", err
+	}
+
+	if dryRun {
+		content := bytes.NewBuffer([]byte{})
+		w := printersinternal.NewPrefixWriter(content)
+		printersinternal.DescribePodTemplate(toTemplate, w)
+		return fmt.Sprintf("will roll back to %s", content.String()), nil
+	}
+
+	// Update DaemonSet template, and retry on conflict
+	skipUpdate := false
+	retryErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var err error
+		ds, err = r.c.Extensions().DaemonSets(ds.Namespace).Get(ds.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if apiequality.Semantic.DeepEqual(toTemplate, &ds.Spec.Template) {
+			skipUpdate = true
+			return nil
+		}
+		ds.Spec.Template = *toTemplate
+		_, err = r.c.Extensions().DaemonSets(ds.Namespace).Update(ds)
```

Comment:
If the resource version isn't the same, `Update` will fail (return conflict error)